### PR TITLE
OptionsResolver test coverage

### PR DESCRIPTION
**Pull request by eventhorizonpl**
Hi,

This PR adds 100% test code coverage to OptionsResolver component.

Best regards,
Michal

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| License | MIT |

### CODE REVIEW

**Comment by Tobion** [comment]
1. it should not be done with the LegacyOptionsResolverTest but the real one
2. private methods should not be tested with reflection but through the public api

**Comment by eventhorizonpl** [comment]
@Tobion 

I introduced the changes you wanted. I have very mixed feelings about the principle number 2. I know that it's good for most cases but sometimes some parts of code are harder to test. I don't see anything wrong with using reflections in such cases. Can this be discussed?

Thanks for review.

**Comment by fabpot** [comment]
I agree with @Tobion, private methods encapsulate logic as black boxes. We need to test them via the regular public API developers are going to use.

**Comment by Tobion** [comment]
@eventhorizonpl I don't see the `true` case (https://github.com/eventhorizonpl/symfony/blob/options_resolver_test_coverage/src/Symfony/Component/OptionsResolver/OptionsResolver.php#L1192) being used in the `OptionsResolver2Dot6Test`. Are you sure it's 100% tests coverage with only the new test class?

**Comment by eventhorizonpl** [comment]
@Tobion 

true case is covered by testResolveFailsIfValueAllowedCallbackReturnsFalse according to a report.

scren

**Comment by Tobion** [comment]
@eventhorizonpl `testResolveFailsIfValueAllowedCallbackReturnsFalse` is only in the legacy test suite. So in master, where the legacy is removed, there would not be 100% test coverage anymore.

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 1]
```diff
@@ -1557,4 +1557,41 @@ public function testCountFailsOutsideResolve()
...
+    /**
+     * @expectedException \Symfony\Component\OptionsResolver\Exception\InvalidOptionsException
+     * @expectedExceptionMessage The option "option" with value array is expected to be of type "string", but is of type "array".
+     */
+    public function testFormatValueArray()
+    {
+        $this->resolver->setDefined('option');
+        $this->resolver->setAllowedTypes('option', 'string');
+
+        $this->resolver->resolve(array('option' => []));
```
Comment (stof) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: please keep the code compatible with PHP 5.3 (this is why Travis is failing)

**Comment by Tobion** [comment]
@eventhorizonpl can you please rewrite the tests to use a dataProvider. This provider (actual type, allowed type, exception message) would then also cover all the other types (from `testResolveFailsIfInvalidType`, `testResolveFailsIfInvalidTypeIsNull`, `testResolveFailsIfNotInstanceOfClass`). This way there is much less repitition.
And you could also do a similar approach to test that passing the correct type succeeds for each type (`testResolveSucceedsIfValidType` etc.).

**Comment by eventhorizonpl** [comment]
@Tobion 

I added dataProvider for my tests and testResolveFailsIfInvalidType, testResolveFailsIfInvalidTypeIsNull, testResolveFailsIfNotInstanceOfClass.

I didn't changed testResolveSucceedsIfValidType* because my current goal is not to fix all tests. Right now it's just about covering code.

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 2]
```diff
@@ -1557,4 +1522,57 @@ public function testCountFailsOutsideResolve()
 
         count($this->resolver);
     }
+
+    /**
+     * @dataProvider getFormatValueData
+     * @expectedException \Symfony\Component\OptionsResolver\Exception\InvalidOptionsException
+     */
+    public function testFormatValue($actualType, $allowedType, $exceptionMessage)
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: please name this test `testResolveFailsIfInvalidType` and at the same lines of the old tests. this makes merging easier.

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 3]
```diff
@@ -1557,4 +1522,57 @@ public function testCountFailsOutsideResolve()
...
+    {
+        $this->resolver->setDefined('option');
+        $this->resolver->setAllowedTypes('option', $allowedType);
+
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: you need to use `setExpectedException` https://phpunit.de/manual/4.1/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions.examples.ExceptionTest3.php to test against the messge as well

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 4]
```diff
@@ -1557,4 +1522,57 @@ public function testCountFailsOutsideResolve()
...
+        $this->resolver->resolve(array('option' => $actualType));
+    }
+
+    public function getFormatValueData()
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: please put this below `testResolveFailsIfInvalidType` and rename it to `provideInvalidTypes`

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 5]
```diff
@@ -1557,4 +1522,57 @@ public function testCountFailsOutsideResolve()
...
+    {
+        return array(
+            array(
+                'actualType' => true,
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: please remove the index keys. they are not needed.

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php (hunk 3 above)
Comment (jakzal) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: Or use the `@expectedExceptionMessage` annotation.

Comment (jakzal) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: Never mind me. I just noticed that the message is dynamic :)

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 6]
```diff
@@ -499,27 +499,29 @@ public function testFailIfSetAllowedTypesFromLazyOption()
     }
 
     /**
+     * @dataProvider provideInvalidTypes
      * @expectedException \Symfony\Component\OptionsResolver\Exception\InvalidOptionsException
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: must be removed

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 7]
```diff
@@ -499,27 +499,29 @@ public function testFailIfSetAllowedTypesFromLazyOption()
...
-     * @expectedExceptionMessage The option "foo" with value 42 is expected to be of type "string", but is of type "integer".
      */
-    public function testResolveFailsIfInvalidType()
+    public function testResolveFailsIfInvalidType($actualType, $allowedType, $exceptionMessage)
     {
-        $this->resolver->setDefined('foo');
-        $this->resolver->setAllowedTypes('foo', 'string');
+        $this->setExpectedException('InvalidArgumentException', $exceptionMessage);
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: `Symfony\Component\OptionsResolver\Exception\InvalidOptionsException`

Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: `setExpectedException` must be called between setAllowedTypes and resolve as the exception is to be expected then.

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 8]
```diff
@@ -499,27 +499,29 @@ public function testFailIfSetAllowedTypesFromLazyOption()
...
-    public function testResolveFailsIfInvalidTypeIsNull()
+    public function provideInvalidTypes()
     {
-        $this->resolver->setDefault('foo', null);
-        $this->resolver->setAllowedTypes('foo', 'string');
-
-        $this->resolver->resolve();
+        return array(
+            array(true, 'string', 'The option "option" with value true is expected to be of type "string", but is of type "boolean".'),
+            array(fopen(__FILE__, "r"), 'string', 'The option "option" with value resource is expected to be of type "string", but is of type "resource".'),
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: we use single quotes `'r'`

File: src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php [hunk 9]
```diff
@@ -499,27 +499,29 @@ public function testFailIfSetAllowedTypesFromLazyOption()
...
-     */
-    public function testResolveFailsIfInvalidTypeIsNull()
+    public function provideInvalidTypes()
     {
-        $this->resolver->setDefault('foo', null);
-        $this->resolver->setAllowedTypes('foo', 'string');
-
-        $this->resolver->resolve();
+        return array(
+            array(true, 'string', 'The option "option" with value true is expected to be of type "string", but is of type "boolean".'),
```
Comment (Tobion) [inline src/Symfony/Component/OptionsResolver/Tests/OptionsResolver2Dot6Test.php]: `false` is missing as test

**Comment by Tobion** [comment]
Thank you @eventhorizonpl.

**Pull request closed without merging**